File: tsSDK/test/test-query.ts
/**
 * Query helper script (markets / orders / positions).
 *
 * Run:
 *   npx tsx test/test-query.ts
 *
 * Env (see test/test-config.ts):
 *   ACCOUNT_ADDRESS (required)
 *   PRIVATE_KEY      (optional; only needed for sending tx)
 *   CELO_RPC_URL / ARBITRUM_RPC_URL (optional)
 *   ORACLE_URL (optional, e.g. "https://api.perpex.ai/prices/")
 *
 * Optional:
 *   CHAIN=celo|arbitrum  (default: celo)
 *   MARKET_ADDRESS     — if set, prints LP (GM) balance for this market
 *   ACCOUNT_ADDRESS    — optional for LP; defaults to SDK account (same as .env ACCOUNT_ADDRESS)
 */

import { abis } from '../src/abis/index'
import type { MarketInfo, MarketsInfoData } from '../src/types/markets'
import { initSdk } from './test-config'
import type { Address } from 'viem'
import { formatUnits } from 'viem'

function getMarketInfoByAddress(
  marketsInfoData: MarketsInfoData | undefined,
  marketAddress: string
): MarketInfo | undefined {
  if (!marketsInfoData) return undefined
  const lower = marketAddress.trim().toLowerCase()
  const key = Object.keys(marketsInfoData).find(
    (k) => k.toLowerCase() === lower
  )
  return key ? marketsInfoData[key] : undefined
}

function selectChain(): 'celo' | 'arbitrum' {
  const raw = (process.env.CHAIN || 'celo').toLowerCase()
  return raw === 'arbitrum' ? 'arbitrum' : 'celo'
}

async function main() {
  const chain = selectChain()
  const sdk = initSdk(chain)

  console.log('\n' + '='.repeat(80))
  console.log(`🔎 Querying SDK state (chain=${chain}, chainId=${sdk.chainId})`)
  console.log(`   account=${sdk.account}`)
  console.log('='.repeat(80) + '\n')

  // 1) Markets list (from local config; then marketInfo requires prices + multicalls)
  const { marketsAddresses, marketsData } = await sdk.markets.getMarkets()

  console.log(`📈 Markets (${marketsAddresses?.length ?? 0})`)
  for (const marketAddress of marketsAddresses || []) {
    const m = marketsData?.[marketAddress]
    console.log(
      `- ${marketAddress}  ${m?.name ?? ''}  ` +
        `(index=${m?.indexTokenAddress ?? ''} long=${
          m?.longTokenAddress ?? ''
        } short=${m?.shortTokenAddress ?? ''})`,
    )
  }
  console.log('')

  // 2) Fetch marketInfo + tokensData once; needed for orders parsing and most position helpers
  const {
    marketsInfoData,
    tokensData,
    pricesUpdatedAt,
  } = await sdk.markets.getMarketsInfo()
  // console.log(
  //   `💱 Prices: updatedAt=${
  //     pricesUpdatedAt ? new Date(pricesUpdatedAt).toISOString() : 'unknown'
  //   } ` +
  //     `(marketsInfoData=${
  //       Object.keys(marketsInfoData || {}).length
  //     }, tokensData=${Object.keys(tokensData || {}).length})`,
  // )
  console.log('')

  // 2b) LP (GM) balance for MARKET_ADDRESS + account (ACCOUNT_ADDRESS or sdk.account)
  const marketAddrEnv = process.env.MARKET_ADDRESS?.trim()
  const accountAddrEnv = process.env.ACCOUNT_ADDRESS?.trim()
  const queryAccount = (accountAddrEnv || sdk.account) as Address

  console.log('💰 LP (GM) balance')
  if (!marketAddrEnv) {
    console.log(
      '   (skip: set MARKET_ADDRESS in .env to query market token balance for an account)',
    )
  } else if (!marketsInfoData) {
    console.log('   (skip: marketsInfoData not available)')
  } else {
    const marketInfo = getMarketInfoByAddress(marketsInfoData, marketAddrEnv)
    if (!marketInfo) {
      console.log(
        `   ❌ Market not found in marketsInfoData: ${marketAddrEnv}`,
      )
    } else {
      const lpToken = marketInfo.marketTokenAddress as Address
      const decimals =
        tokensData?.[marketInfo.marketTokenAddress]?.decimals ?? 18
      const raw = (await sdk.publicClient.readContract({
        address: lpToken,
        abi: abis.ERC20 as any,
        functionName: 'balanceOf',
        args: [queryAccount],
      })) as bigint
      const human = formatUnits(raw, decimals)
      console.log(`   market:    ${marketAddrEnv}`)
      console.log(`   LP token:  ${lpToken}`)
      console.log(`   account:   ${queryAccount}`)
      console.log(`   balance:   ${raw.toString()} (${human} GM, decimals=${decimals})`)
    }
  }
  console.log('')

  // 3) Orders (current account)
  if (!marketsInfoData || !tokensData) {
    console.log('🧾 Orders: skipped (missing marketsInfoData/tokensData)')
  } else {
    const { count, ordersInfoData } = await sdk.orders.getOrders({
      marketsInfoData,
      tokensData,
    })

    console.log(`🧾 Orders (${count})`)
    for (const [orderKey, order] of Object.entries(ordersInfoData)) {
      // Keep output robust across SDK type changes.
      const market =
        (order as any).marketAddress ||
        (order as any).marketInfo?.marketTokenAddress
      const orderType = (order as any).orderType
      const isLong = (order as any).isLong
      const sizeUsd = (order as any).sizeDeltaUsd ?? (order as any).sizeInUsd
      console.log(
        `- key=${orderKey} type=${orderType} isLong=${isLong} market=${market} sizeUsd=${String(
          sizeUsd,
        )}`,
      )
    }
  }
  console.log('')

  // 4) Positions (raw positions from chain; does not require prices for fallback path)
  if (!marketsData || !tokensData) {
    console.log('📌 Positions: skipped (missing marketsData/tokensData)')
  } else {
    const { positionsData } = await sdk.positions.getPositions({
      marketsData,
      tokensData,
      start: 0,
      end: 1000,
    })

    const keys = Object.keys(positionsData || {})
    console.log(`📌 Positions (${keys.length})`)
    for (const key of keys) {
      const p = (positionsData as any)[key]
      console.log(
        `- key=${key} market=${p.marketAddress} collateral=${p.collateralTokenAddress} ` +
          `isLong=${p.isLong} sizeUsd=${String(
            p.sizeInUsd,
          )} sizeTokens=${String(p.sizeInTokens)} collateralAmount=${String(
            p.collateralAmount,
          )}`,
      )
    }
  }

  console.log('\n✅ Done.\n')
}

main().catch((e) => {
  console.error('\n❌ Query failed:')
  console.error(e?.stack || e?.message || e)
  process.exit(1)
})

File: pythonSDK/scripts/test_query.py
import os
from datetime import datetime, timezone
from typing import Optional

import json

from web3 import Web3

from updown_sdk import UpdownSdk
from updown_sdk.abis import ERC20_ABI
from updown_sdk.env import load_sdk_dotenv


def _iso(ms: Optional[int]) -> str:
    if not ms:
        return "unknown"
    return datetime.fromtimestamp(ms / 1000, tz=timezone.utc).isoformat().replace("+00:00", "Z")


def _select_chain() -> str:
    raw = (os.getenv("CHAIN") or "celo").lower()
    return "arbitrum" if raw == "arbitrum" else "celo"


def _chain_id(chain: str) -> int:
    return 42161 if chain == "arbitrum" else 42220


def _rpc_url(chain: str) -> str:
    if chain == "arbitrum":
        return os.getenv("ARBITRUM_RPC_URL", "https://arb1.arbitrum.io/rpc")
    return os.getenv("CELO_RPC_URL", "https://forno.celo.org")

def _token_decimals_from_config(chain_id: int, token_address: str) -> int:
    """Decimals for any token in config (incl. market/GM token); mirrors TS tokens.json lookup."""
    import updown_sdk as pkg

    cfg_path = os.path.join(os.path.dirname(pkg.__file__), "config", "tokens.json")
    try:
        with open(cfg_path, "r", encoding="utf-8") as f:
            data = json.load(f)
        for t in data.get(str(chain_id), []):
            if str(t.get("address", "")).lower() == token_address.lower():
                return int(t.get("decimals") or 18)
    except Exception:
        pass
    return 18


def _format_token_amount(raw: int, decimals: int) -> str:
    if decimals <= 0:
        return str(raw)
    neg = raw < 0
    n = abs(int(raw))
    scale = 10**decimals
    whole = n // scale
    frac = n % scale
    s = f"{whole}.{str(frac).zfill(decimals)}".rstrip("0").rstrip(".")
    return ("-" if neg else "") + (s or "0")


def _print_lp_balance(*, w3: Web3, chain_id: int, account: str, market_token_address: str, markets) -> None:
    """
    User LP = ERC20 balance of the market (GM) token — same as TS `operations.withdraw` balanceOf check.
    `market_token_address` is `marketTokenAddress` from markets config (not index token).
    """
    acct = Web3.to_checksum_address(account.strip())
    mt = Web3.to_checksum_address(market_token_address.strip())
    token = w3.eth.contract(address=mt, abi=ERC20_ABI)
    raw = int(token.functions.balanceOf(acct).call())
    dec = _token_decimals_from_config(chain_id, mt)
    human = _format_token_amount(raw, dec)
    name = ""
    for m in markets:
        if m.market_token_address.lower() == mt.lower():
            sym_by = _token_symbol_map(chain_id)
            idx_sym = sym_by.get(m.index_token_address.lower(), "?")
            short_sym = sym_by.get(m.short_token_address.lower(), "?")
            name = (m.name or "").strip() or f"{idx_sym}/{short_sym}"
            break
    print("🏦 LP (GM) balance")
    print(f"   account:   {acct}")
    print(f"   market:    {mt}")
    if name:
        print(f"   pair/name: {name}")
    print(f"   decimals:  {dec}")
    print(f"   balance:   {human}  (raw={raw})")
    print("")


def _token_symbol_map(chain_id: int) -> dict[str, str]:
    """
    TS prints market names from config; python markets.json may omit `name`.
    Build a best-effort address->symbol map from pythonSDK config/tokens.json
    without requiring oracle prices.
    """
    import updown_sdk as pkg

    cfg_path = os.path.join(os.path.dirname(pkg.__file__), "config", "tokens.json")
    try:
        with open(cfg_path, "r", encoding="utf-8") as f:
            data = json.load(f)
        items = data.get(str(chain_id), [])
        out: dict[str, str] = {}
        for t in items:
            addr = str(t.get("address") or "").lower()
            sym = str(t.get("symbol") or "")
            if addr and sym:
                out[addr] = sym
        return out
    except Exception:
        return {}


def main():
    load_sdk_dotenv()
    chain = _select_chain()
    sdk = UpdownSdk(
        chain_id=_chain_id(chain),
        rpc_url=_rpc_url(chain),
        oracle_url=os.getenv("ORACLE_URL", "https://api.perpex.ai/prices/"),
        subsquid_url=os.getenv("SUBSQUID_URL") or os.getenv("SUBGRAPH_URL"),
        account=os.getenv("ACCOUNT_ADDRESS"),
    )

    print("\n" + "=" * 80)
    print(f"🔎 Querying SDK state (chain={chain}, chainId={sdk.chain_id})")
    print(f"   account={sdk.account}")
    print("=" * 80 + "\n")

    markets = sdk.markets.get_markets()
    print(f"📈 Markets ({len(markets)})")
    sym_by_addr = _token_symbol_map(sdk.chain_id)
    for m in markets:
        name = (m.name or "").strip()
        if not name:
            idx_sym = sym_by_addr.get(m.index_token_address.lower(), "UNKNOWN")
            short_sym = sym_by_addr.get(m.short_token_address.lower(), "UNKNOWN")
            name = f"{idx_sym}/{short_sym}"
        print(
            f"- {m.market_token_address}  {name}  "
            f"(index={m.index_token_address} long={m.long_token_address} short={m.short_token_address})"
        )
    print("")

    # tokensData/pricesUpdatedAt (best-effort)
    try:
        tokens_data, prices_updated_at = sdk.tokens.get_tokens_data()
        # Minimal marketsInfoData count: markets that have prices for index/long/short tokens
        priced = set(a.lower() for a in tokens_data.keys())
        markets_info_count = 0
        for m in markets:
            if (
                m.index_token_address.lower() in priced
                and m.long_token_address.lower() in priced
                and m.short_token_address.lower() in priced
            ):
                markets_info_count += 1
        print(
            f"💱 Prices: updatedAt={_iso(prices_updated_at)} "
            f"(marketsInfoData={markets_info_count}, tokensData={len(tokens_data)})"
        )
    except Exception as e:
        print(f"💱 Prices: updatedAt=unknown (marketsInfoData=0, tokensData=0) error={e}")
    print("")

    # Orders
    if not sdk.account:
        print("🧾 Orders: skipped (missing account)\n")
    else:
        try:
            orders = sdk.orders.get_orders()
            print(f"🧾 Orders ({len(orders)})")
            for order_key, o in orders.items():
                print(
                    f"- key={order_key} type={o.order_type} isLong={o.is_long} "
                    f"market={o.market_address} sizeUsd={o.size_delta_usd}"
                )
        except Exception as e:
            print(f"🧾 Orders: failed ({e})")
        print("")

    # Positions
    if not sdk.account:
        print("📌 Positions: skipped (missing account)\n")
    else:
        try:
            markets_dump = sdk.markets.dump_markets()
            # Mirror TS debug: whether we can build marketsKeys with prices
            market_keys, _market_prices = sdk.positions._build_market_prices(markets_dump)  # type: ignore[attr-defined]
            if len(market_keys) == 0:
                print(
                    "⚠️  No markets with prices found. marketsKeys is empty. "
                    "This means oracle tickers did not cover required tokens. "
                    "Falling back to getAccountPositions (doesn't require prices)."
                )
            else:
                print(f"✅ Found {len(market_keys)} markets with prices. Using getAccountPositionInfoList.")

            positions = sdk.positions.get_positions(markets=markets_dump)
            print(f"📌 Positions ({len(positions)})")
            for key, p in positions.items():
                print(
                    f"- key={key} market={p.market} collateral={p.collateral_token} "
                    f"isLong={p.is_long} sizeUsd={p.size_in_usd} sizeTokens={p.size_in_tokens} collateralAmount={p.collateral_amount}"
                )
        except Exception as e:
            print(f"📌 Positions: failed ({e})")

    # LP = market (GM) token balance for account + market address (TS: ERC20 balanceOf on marketTokenAddress)
    lp_market = (os.getenv("MARKET_ADDRESS") or os.getenv("MARKET_TOKEN_ADDRESS") or "").strip()
    lp_account = (os.getenv("ACCOUNT_ADDRESS") or (sdk.account or "") or "").strip()
    if lp_market and lp_account:
        try:
            _print_lp_balance(
                w3=sdk.w3,
                chain_id=sdk.chain_id,
                account=lp_account,
                market_token_address=lp_market,
                markets=markets,
            )
        except Exception as e:
            print(f"🏦 LP (GM) balance: failed ({e})\n")
    elif lp_market and not lp_account:
        print("🏦 LP (GM) balance: skipped (set ACCOUNT_ADDRESS or pass account to UpdownSdk)\n")

    print("\n✅ Done.\n")


if __name__ == "__main__":
    main()

